numba_cuda/numba/cuda/cudadrv/driver.py

```diff
@@ -2406,6 +2416,49 @@ def get_function(self, name):
     def get_global_symbol(self, name):
         """Return a MemoryPointer referring to the named symbol"""
 
+    def _set_callables(self, callbacks, attr):
+        if not isinstance(callbacks, list):
+            raise TypeError("callbacks must be a list")
+
+        for f in callbacks:
+            if not callable(f):
+                raise TypeError("callback must be callable")
+
+        setattr(self, attr, callbacks)
+
+    def set_setup_functions(self, callbacks):
+        """Set the setup callback for this module"""
+        self._set_callables(callbacks, "setup_functions")
+
+    def set_teardown_functions(self, callbacks):
+        """Set the finalize callback for this module"""
+        self._set_callables(callbacks, "teardown_functions")
+
+    def setup(self, stream):
+        """Call the setup functions for cumodule with the given `stream`"""
+        if self.setup_functions is None:
+            return
+
+        for f in self.setup_functions:
+            f(weakref.proxy(self), stream)
```

Comment:
Is the `weakref.proxy` used because we expect the setup function might retain a reference to this module? Do we expect users to do that? (e.g. is that a requirement in the use cases you've been discussing with potential users of this?)

Reply:
I guess I'm thinking - is this accidentally just mirroring the `set_finalizers()` function, where we use `weakref.proxy()` to avoid creating a cycle that would prevent us from every unloading the module?